[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113209
* __->__ #120454

With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

fegin committed:
[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd

With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

[ghstack-poisoned]

fegin committed:
Update on "[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd"


With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

yf225 commented on benchmarks/dynamo/common.py:
```diff
@@ -2640,10 +2646,15 @@ def warmup(fn, model, example_inputs, mode, niters=5):
+        model = self.deepcopy_and_maybe_parallelize(orig_model)
+        if experiment.func is speedup_experiment:
+            # If DDP + compiler is enabled, we need to use a different
+            compiled_model = self.deepcopy_and_maybe_parallelize(orig_model)
```
In the future for large models, I suspect we might need to `del orig_model` after deepcopy to get more memory and avoid OOM

xmfan commented on benchmarks/dynamo/common.py:
```diff
@@ -737,11 +738,16 @@ def maybe_mark_profile(*args, **kwargs):
+        and torch.distributed.is_initialized()
+        and torch.distributed.get_rank() == 0
+    ):
+        output_csv(
```
we still need to output_csv for single gpu, this gonna break CI dashboard

fegin replied:
oh, right, good catch, let me change the check.

xmfan commented:
shouldn't this also affect accuracy bench?

xmfan commented on benchmarks/dynamo/common.py:
```diff
@@ -737,11 +738,16 @@ def maybe_mark_profile(*args, **kwargs):
-        row,
-    )
+    if (
+        torch.distributed.is_available()
+        and torch.distributed.is_initialized()
+        and torch.distributed.get_rank() == 0
```
use `args.init_distributed` to check first, then check torch.distributed submodules

otherwise, all benchmarks will require distributed build

fegin replied:
It should not break the single GPU case. If distributed is not built, `torch.distributed.is_available()` returns False and it is the very first check.

fegin committed:
Update on "[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd"


With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd"


With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[PT2D] Make the speedup benchmark works with DDP + CompiledAutograd"


With DDP + CompiledAutograd, we could not use the same parallelized model to do the test. This PR copies the model.

Differential Revision: [D54094257](https://our.internmc.facebook.com/intern/diff/D54094257/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

desertfire commented:
@pytorchbot revert -m "breaks nightly dashboard cudagraphs run" -c nosignal

desertfire commented:
I think this PR missed some `model`->`compiled_model` changes.
